packages/fxa-event-broker/lib/notificationProcessor.ts

```diff
@@ -0,0 +1,110 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { SQS } from 'aws-sdk';
+import { Consumer } from 'sqs-consumer';
+import * as Joi from 'typesafe-joi';
+
+import { Logger } from 'mozlog';
+import { Datastore } from './db';
+
+// Event strings
+const LOGIN_EVENT = 'login';
+const SUBSCRIPTION_UPDATE_EVENT = 'subscription:update';
+
+// Message schemas
+const BASE_MESSAGE_SCHEMA = Joi.object()
+  .keys({
+    event: Joi.string().required(),
+    uid: Joi.string().required()
+  })
+  .unknown(true)
+  .required();
+
+const LOGIN_SCHEMA = Joi.object()
+  .keys({
+    clientId: Joi.string().required(),
+    deviceCount: Joi.number()
+      .integer()
+      .required(),
+    email: Joi.string().required(),
+    event: Joi.string()
+      .valid(LOGIN_EVENT)
+      .required(),
+    service: Joi.string().required(),
+    uid: Joi.string().required(),
+    userAgent: Joi.string().required()
+  })
+  .unknown(true)
+  .required();
+
+const SUBSCRIPTION_UPDATE_SCHEMA = Joi.object()
+  .keys({
+    event: Joi.string()
+      .valid(SUBSCRIPTION_UPDATE_EVENT)
+      .required(),
+    isActive: Joi.bool().required(),
+    productCapabilities: Joi.array()
+      .items(Joi.string())
+      .required(),
+    productName: Joi.string().required(),
+    subscriptionId: Joi.string().required(),
+    uid: Joi.string().required()
+  })
+  .unknown(true)
+  .required();
+
+class ServiceNotificationProcessor {
+  public readonly app: Consumer;
+  private readonly db: Datastore;
+  private readonly logger: Logger;
+
+  constructor(logger: Logger, db: Datastore, queueUrl: string, sqs: SQS) {
+    this.db = db;
+    this.logger = logger;
+    this.app = Consumer.create({
+      handleMessage: async (message: SQS.Message) => {
+        return await this.handleMessage(message);
+      },
+      queueUrl,
+      sqs
+    });
+    this.app.on('error', err => {
+      logger.error('consumerError', { err });
+    });
+
+    this.app.on('processing_error', err => {
+      logger.error('processingError', { err });
+    });
+  }
+
+  public start() {
+    this.app.start();
+  }
+
+  public stop() {
+    this.app.stop();
+  }
+
+  private async handleMessage(sqsMessage: SQS.Message) {
+    const body = JSON.parse(sqsMessage.Body || '{}');
+    const message = Joi.attempt(JSON.parse(body.Message), BASE_MESSAGE_SCHEMA);
+    switch (message.event) {
+      case LOGIN_EVENT: {
+        const loginMessage = Joi.attempt(message, LOGIN_SCHEMA);
+        await this.db.storeLogin(loginMessage.uid, loginMessage.clientId);
+        return;
+      }
+      case SUBSCRIPTION_UPDATE_EVENT: {
+        const subMessage = Joi.attempt(message, SUBSCRIPTION_UPDATE_SCHEMA);
+        const clientIds = await this.db.fetchClientIds(subMessage.uid);
+        // TODO: Queue a subscription event for each clientId for delivery.
+        return;
+      }
+    }
+    // Anything that isn't a message we want
+    this.logger.debug('unwantedMessage', { message });
```

Comment:
should this remain `debug` or be a `warn` or `info` so we can catch an unwanted message in prod?

Reply:
Since we are listening for all events, and only processing the login/subscription ones (this will hang off the main service notification SNS topic), there will be lots of messages we don't want. I'll check with ops to see if they'd prefer to try and configure this SQS listener to filter it down to just the two message event types we want. Later on we will want all of them though, for event logging.